src/main/java/build/buildfarm/instance/shard/ShardInstance.java

```diff
@@ -418,6 +426,26 @@ public void start() {
     if (operationQueuer != null) {
       operationQueuer.start();
     }
+
+    if (config.getRunIndexer()) {
+      indexerService.scheduleAtFixedRate(
+          () -> {
+            try {
+
+              // ensure only one scheduler performs the re-indexing.
+              if (backplane.takeIndexerLock()) {
+                backplane.runIndexer();
```

Comment:
There are two failure cases not considered here: the first is easy - once the lock is acquired, a finally should ensure that even if runIndexer() throws an exception, that the indexer lock is released.

The second is the abnormal termination of a process, which means that there needs to be an upper bound on the lock acquisition time. Ideally that should be tied to the lifetime of some redis connection, but in a cluster, that will be difficult. Preferentially there should be a) a required poll of the activity in the indexer to extend the lock so that the timeout can remain short, and b) some clear indication which node is performing an index currently, and queryability as to its progress